### PR TITLE
POL-164 - Update Azure policies with credential service

### PR DESCRIPTION
**Pull request by praut1**
### Description

Modified Azure policies to support credential service

### Issues Resolved

[List any existing issues this PR resolves]

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

Running Template links and screenshots

Azure Idle Compute Instances
https://governance.rightscale.com/org/1105/projects/60073/policy-incidents/5e255b2632401200016e0b0f
image

Azure Blob Storage Optimization
https://governance.rightscale.com/org/1105/projects/60073/policy-incidents/5e2942f74ec7b300014e6b1a
image

Azure Publicly Accessible SQL Managed Instances
https://governance.rightscale.com/org/1105/projects/60073/policy-incidents/5e297b574ec7b300014e6b6e
image

Azure Delete Unused SQL Database
https://governance.rightscale.com/org/1105/projects/60073/policy-incidents/5e26d511cbdfbd000104bda2
image

Above 4 policies have been tested successfully.
Azure Disallowed Regions - Separate ticket has been created to resolve delete resources issue POL-182, hence the policy has not been included in this pull request.
Azure Instances Utilization from Log Analytics - This policy has been already merged to master with the credential service changes in POL-101.
Azure Publicly Accessible Managed SQL Instance - Positive scenario is working fine, hence adding the policy to this pull request but created separate ticket to modify the policy to achieve empty tag list scenario, POL-181.

### CODE REVIEW

File: cost/azure/idle_compute_instances/README.md [hunk 1]
```diff
@@ -12,28 +12,12 @@ The policy leverages the Azure API to check all instances and then checks the in
...
-1. Follow steps to [Create an Azure Active Directory Application](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal#create-an-azure-active-directory-application)
-1. Grant the Azure AD Application access to the necessary subscription(s)
-1. [Retrieve the Application ID & Authentication Key](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal#get-application-id-and-authentication-key)
-1. Create RightScale Credentials with values that match the Application ID (Credential name: `AZURE_APPLICATION_ID`) & Authentication Key (Credential name: `AZURE_APPLICATION_KEY`)
-1. [Retrieve your Tenant ID](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal#get-tenant-id)
-
-## Required RightScale Roles
-
-- `credential_viewer`
+This policy requires the Azure Resource Manager Credential. When applying the policy select the appropriate credentials from the list for your tenant. If such credential doesn't exist please contact your cloud admin to create the Credential.
```
Comment (cdwilhelm) [inline cost/azure/idle_compute_instances/README.md]: See the readme_guidelines for the correct Prerequisite text

File: cost/azure/object_storage_optimization/README.md [hunk 2]
```diff
@@ -1,57 +1,37 @@
...
-
-### Installation
+## Prerequesites
 
-1. Follow steps to [Create an Azure Active Directory Application](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal#create-an-azure-active-directory-application)
-1. Grant the Azure AD Application access to the necessary subscription(s)
-1. [Retrieve the Application ID & Authentication Key](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal#get-application-id-and-authentication-key)
-1. Create RightScale Credentials with values that match the Application ID (Credential name: `AZURE_APPLICATION_ID`) & Authentication Key (Credential name: `AZURE_APPLICATION_KEY`)
-1. [Retrieve your Tenant ID](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal#get-tenant-id)
+This policy requires the Azure Resource Manager Credential. When applying the policy select the appropriate credentials from the list for your tenant. If such credential doesn't exist please contact your cloud admin to create the Credential.
```
Comment (cdwilhelm) [inline cost/azure/object_storage_optimization/README.md]: see readme_guideline for prerequisite text

File: cost/azure/unused_sql_databases/README.md [hunk 3]
```diff
@@ -1,60 +1,40 @@
...
-### Prerequisites
+## Prerequisites
 
-- Azure Service Principal (AKA Azure Active Directory Application) with the appropriate permissions to manage resources in the target subscription
-- The following RightScale Credentials
-  - `AZURE_APPLICATION_ID`
-  - `AZURE_APPLICATION_KEY`
-  
-### Installation
+This policy requires the Azure Resource Manager Credential. When applying the policy select the appropriate credentials from the list for your tenant. If such credential doesn't exist please contact your cloud admin to create the Credential.
```
Comment (cdwilhelm) [inline cost/azure/unused_sql_databases/README.md]: same as other readmes

File: security/azure/sql_publicly_accessible_managed_instance/README.md [hunk 4]
```diff
@@ -1,62 +1,46 @@
...
 
-### Installation
+## Prerequesites
 
-1. Follow steps to [Create an Azure Active Directory Application](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal#create-an-azure-active-directory-application)
-1. Grant the Azure AD Application access to the necessary subscription(s)
-1. [Retrieve the Application ID & Authentication Key](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal#get-application-id-and-authentication-key)
-1. Create RightScale Credentials with values that match the Application ID (Credential name: `AZURE_APPLICATION_ID`) & Authentication Key (Credential name: `AZURE_APPLICATION_KEY`)
-1. [Retrieve your Tenant ID](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal#get-tenant-id)
+This policy requires the Azure Resource Manager Credential. When applying the policy select the appropriate credentials from the list for your tenant. If such credential doesn't exist please contact your cloud admin to create the Credential.
```
Comment (cdwilhelm) [inline security/azure/sql_publicly_accessible_managed_instance/README.md]: see another readme for prerequisite example

Comment (praut1) [reply security/azure/sql_publicly_accessible_managed_instance/README.md]: Earlier referred POL-151 for Prerequisites text, now changed it by referring to the policies which have been merged to master recently viz. Azure Rightsize SQL Databases and Azure Inefficient Instance Utilization using Log Analytics.

File: cost/azure/unused_sql_databases/README.md (hunk 3 above)
Comment (praut1) [reply cost/azure/unused_sql_databases/README.md]: Earlier referred POL-151 for Prerequisites text, now changed it by referring to the policies which have been merged to master recently viz. Azure Rightsize SQL Databases and Azure Inefficient Instance Utilization using Log Analytics.

File: cost/azure/object_storage_optimization/README.md (hunk 2 above)
Comment (praut1) [reply cost/azure/object_storage_optimization/README.md]: Earlier referred POL-151 for Prerequisites text, now changed it by referring to the policies which have been merged to master recently viz. Azure Rightsize SQL Databases and Azure Inefficient Instance Utilization using Log Analytics.

File: cost/azure/idle_compute_instances/README.md (hunk 1 above)
Comment (praut1) [reply cost/azure/idle_compute_instances/README.md]: Earlier referred POL-151 for Prerequisites text, now changed it by referring to the policies which have been merged to master recently viz. Azure Rightsize SQL Databases and Azure Inefficient Instance Utilization using Log Analytics.

**Comment by praut1** [comment]
Hi @cdwilhelm, prerequisite changes have been done yesterday, please have a look at it once and let me know if anything else needs to be changed.
Also, do we need to apply sorting logic to any of the policies before merging them to the master?

**Comment by cdwilhelm** [comment]
@praut1   Just a minor change on the readme's.  please remove the ## Azure Required Permissions Header and move the permissions bullets under the Cloud Configuration header in the Prerequisite section.

File: security/azure/sql_publicly_accessible_managed_instance/check_for_publicly_accessible_azure_sql_managed_instance.pt [hunk 5]
```diff
@@ -116,12 +116,57 @@ datasource "ds_list_sql_managed_instances" do
...
+        if(typeof(tags) != "undefined" && tags!=null){
+          for(var k=0;k<tags.length;k++){
+            var tag=tags[k];
+              if(param_exclusion_tag_key_lowercase.indexOf((tag['tagKey']).toLowerCase()) !== -1){
+               is_tag_matched = true;
+              }
+          }
+        }         
+        if(!is_tag_matched){
+          res.push({
```
Comment (cdwilhelm) [inline security/azure/sql_publicly_accessible_managed_instance/check_for_publicly_accessible_azure_sql_managed_instance.pt]: please sort by subscriptionName and region/location

File: cost/azure/idle_compute_instances/azure_idle_compute_instances.pt [hunk 6]
```diff
@@ -195,10 +205,10 @@ policy "azure_resource_policy" do
     detail_template <<-EOS
 # Azure idle compute instances
 
-| Region/Location | Name | CPU Average | Resource Group | Tags |
-| -------- | ---- | ----------- | -------------- | ---- |
+| Subscription Name | Region/Location | Name | CPU Average | Resource Group | Tags |
```
Comment (cdwilhelm) [inline cost/azure/idle_compute_instances/azure_idle_compute_instances.pt]: please sort by Subscription name, then region

File: cost/azure/object_storage_optimization/azure_object_storage_optimization.pt [hunk 7]
```diff
@@ -1,10 +1,16 @@
 name "Azure Blob Storage Optimization"
```
Comment (cdwilhelm) [inline cost/azure/object_storage_optimization/azure_object_storage_optimization.pt]: Please add sorting by the blobname

File: cost/azure/unused_sql_databases/azure_unused_sql_databases.pt [hunk 8]
```diff
@@ -223,10 +234,10 @@ policy "policy_azure_db_utilization" do
     detail_template <<-EOS
 # Azure Unused SQL Databases
 
-| Region | Name | Resource Group | Kind | Type | SKU - Name | SKU - Tier | SKU - Capacity |
-| ------ | ---- | -------------- | ---- | ---- | -----------| ---------- | -------------- |
+| Subscription Name | Region | Name | Resource Group | Kind | Type | SKU - Name | SKU - Tier | SKU - Capacity |
```
Comment (cdwilhelm) [inline cost/azure/unused_sql_databases/azure_unused_sql_databases.pt]: please add sorting by subscription name and region

File: security/azure/sql_publicly_accessible_managed_instance/check_for_publicly_accessible_azure_sql_managed_instance.pt (hunk 5 above)
Comment (praut1) [reply security/azure/sql_publicly_accessible_managed_instance/check_for_publicly_accessible_azure_sql_managed_instance.pt]: Sorted by subscriptionName and location

File: cost/azure/idle_compute_instances/azure_idle_compute_instances.pt (hunk 6 above)
Comment (praut1) [reply cost/azure/idle_compute_instances/azure_idle_compute_instances.pt]: Sorted by subscriptionName and location

File: cost/azure/object_storage_optimization/azure_object_storage_optimization.pt (hunk 7 above)
Comment (praut1) [reply cost/azure/object_storage_optimization/azure_object_storage_optimization.pt]: Sorted by blobname

File: cost/azure/unused_sql_databases/azure_unused_sql_databases.pt (hunk 8 above)
Comment (praut1) [reply cost/azure/unused_sql_databases/azure_unused_sql_databases.pt]: Sorted by subscriptionName and location. Also, changed the field name from Region to Location in reporting part as we are fetching the value by .location

**Pull request merged**